src/sentry/ratelimits/sliding_windows.py

```diff
@@ -0,0 +1,209 @@
+    requested: int
+
+    # Which quotas to check against. The requested amount must "fit" into all
+    # quotas.
+    quotas: Sequence[Quota]
+
+
+@dataclass(frozen=True)
+class GrantedQuota:
+    # The prefix from RequestedQuota
+    prefix: str
+
+    # How much of RequestedQuota.requested can actually be used.
+    granted: int
+
+
+class SlidingWindowRateLimiter(Service):
+    def __init__(self, **options: Any) -> None:
+        pass
+
+    def check_within_quotas(
+        self, requests: Sequence[RequestedQuota], timestamp: Optional[int] = None
+    ) -> Tuple[int, Sequence[GrantedQuota]]:
+        """
+        Given a set of quotas requests and limits, compute how much quota could
+        be consumed.
+        """
+        raise NotImplementedError()
+
+    def use_quotas(
+        self, requests: Sequence[RequestedQuota], grants: Sequence[GrantedQuota], timestamp: int
+    ) -> None:
+        """
+        Given a set of requests and the corresponding return values from
+        `check_within_quotas`, consume the quotas.
+
+        This is a separate method because 1) the redis backend is not strongly
+        consistent anyway 2) the metrics string indexer may not want to
+        increment quotas if it crashes halfway through the database write. In
+        other words, it may want to first check quotas, execute the writes,
+        then apply the quotas.
+        """
+        raise NotImplementedError()
+
+    def check_and_use_quotas(
+        self, requests: Sequence[RequestedQuota], timestamp: Optional[int] = None
+    ) -> Sequence[GrantedQuota]:
+        """
+        Check the quota requests in Redis and consume the quota in one go.
+        """
+        timestamp, grants = self.check_within_quotas(requests, timestamp)
+        self.use_quotas(requests, grants, timestamp)
+        return grants
+
+
+class RedisSlidingWindowRateLimiter(SlidingWindowRateLimiter):
+    def __init__(self, **options: Any) -> None:
+        cluster_key = options.get("cluster", "default")
+        self.client = redis.redis_clusters.get(cluster_key)
+        super().__init__(**options)
+
+    def validate(self) -> None:
+        try:
+            self.client.ping()
+        except Exception as e:
+            raise InvalidConfiguration(str(e))
+
+    def _build_redis_key_raw(self, prefix: str, window: int, granularity: int, granule: int) -> str:
+        return f"sliding-window-rate-limit:{prefix}:{window}:{granularity}:{granule}"
+
+    def _build_redis_key(self, request: RequestedQuota, quota: Quota, granule: int) -> str:
+        return self._build_redis_key_raw(
+            prefix=quota.prefix_override or request.prefix,
+            window=quota.window,
+            granularity=quota.granularity,
+            granule=granule,
+        )
+
+    def check_within_quotas(
+        self, requests: Sequence[RequestedQuota], timestamp: Optional[int] = None
+    ) -> Tuple[int, Sequence[GrantedQuota]]:
+        if timestamp is None:
```

Comment:
this function is a little difficult to follow along with. The code is totally fine but it's hard to map what is going on in the code to what is going on in redis. 

An example of how the data can look like in the comments can make it much easier to follow along with what is happening. This was done for the [snuba rate limiter](https://github.com/getsentry/snuba/blob/master/snuba/state/rate_limit.py#L143-L170).

You can also provide the intention behind code in [human language](https://github.com/getsentry/sentry/blob/master/src/sentry/scripts/ratelimits/api_limiter.lua#L39-L40) so that the intent is clear without the reader having to run the code in their head to figure it out.